README.md
# PersistentStorage

[![Build Status](https://travis-ci.org/joelbyler/persistent_storage.svg?branch=master)](https://travis-ci.org/joelbyler/persistent_storage)

Stores and retrieves terms from small flat files on embedded systems.

`PersistentStorage` is intended for trivial persistent storage of basic system and application configuration information on embedded systems.   It is not intended to be a replacement for dets, sqlite, or many other far more capable databases.  It favors simplicity and robustness over performance and capability.

**IMPORTANT -- INCOMPATIBLE API CHANGE**

The API has has changed significantly for PersistentStorage 0.10.x -- you can find a summary of the changes in the [[CHANGELOG]].

If your project requires compatibility with the old (now deprecated) api, please make sure you specify "~> 0.9.0" in your project dependencies.

## Philosophy

PersistentStorage manages one more more persistent _tables_, each identified by an atom.  Each table maps to a user-configured directory in the filesystem.   The table keys and associated directories are configured in the application environment using `config.exs`.

Each table persists a set of _keys_ and associated values, where each key maps to a flat file in the table's directory that holds the serialized term.

The simple one-file-per-key/value model favors robustness over speed, reducing risk of write corruption on embedded devices that may power down unexpectedly.

Writes always write and close a file, so are very slow (but plenty fast enough for the kind of device configuration data that doesn't churn frequently).

Reads of a key from disk are slow the first time (they open and read a file), but the term is subsequently cached in ets, so further reads of the same key in a storage area are very fast, and no application-level cacheing is needed.

## Installation

1. Add `persistent_storage` to your list of dependencies in `mix.exs`:

  ```elixir
  def deps do
    [{:persistent_storage, "~> 0.10.0"}]
  end
  ```

2. Ensure `persistent_storage` is started before your application:

  ```elixir
  def application do
    [applications: [:persistent_storage]]
  end
  ```

## Usage

Let's say your device wants a couple kinds of non-volatile storage.  

The first kind, which we'll call `settings`, will be used for storing user
settings. We're going to put it on the standard Nerves application data volume, `/root`.

We're going to assume that the /root partition might be rewritten anytime the user does something "resets to factory settings".

The second kind, which we'll call `provisioning`, is more persistent, and is perhaps on a volume usually mounted read-only (like /boot), that persists even when the application volume gets reformatted.

#### Configuration

Define one or more tables in your config.exs as follows...

```elixir
# my_app/config/config.exs
...
config :persistent_storage, tables: [
  settings: [path: "/root/storage/settings"],
  provisioning: [path: "/boot/provisioning"]
]
```

#### Writing

Writing to setttings, assuming the configuration in the previous section.  This will create the file at /root/storage/settings/network.

```elixir
iex> PersistentStorage.put :settings, :network, %{
  ip_address: {192,168,1,100}, mode: :static
}
```

#### Reading

Assuming we wrote as above, we can read it (this reads from ets cache if possible)...

```elixir
iex> PersistentStorage.get :settings, :network
%{ip_address: {192,168,1,100}, mode: :static}
```
Read some provisioning data -- for purposes of this example, we
assume it was written when device  was first flashed)

```elixir
iex> PersistentStorage.get :provisioning, :device_data
[serial_number: "302F1010", mfg_timestamp: "2016-05-04T03:28:35.279977Z"]
```


CHANGELOG.md
# CHANGELOG

## v0.10.1 (16-Feb-2017)

- removed inappropriate debugging IO.puts that was left in code
- added a dedicated process to own the ets tables
- added warnings

## v0.10.0 (15-Feb-2017)

__IMPORTANT!!__  -- The API has changed a fair in an incompatible way since v0.9.x. It also has a new home (for now) at ghitchens/persistent_storage

Reference cellulose/persistent_storage and specify version ~>0.9.0 in your deps if you have code compiled against the old API.

* Backwards incompatible changes
  - now configured in `config.exs` not in `setup()`
  - APIs now take a storage as first parameter
  - setting of multiple keys in a single put is no longer supported (this was not used heavily and complicated the API)
  - put(), delete() return better responses, but might break previous matches

* Other changes
  - now using Apache-2.0 LICENSE

## v0.9.0 (2016)

- Broken out from http://github.com:/ghitchens/echo


config/test.exs
use Mix.Config

config :persistent_storage, tables: [
  test1: [ path: "/tmp/persistent_storage_test_1" ],
  test2: [ path: "/tmp/persistent_storage_test_2" ]
]


test/persistent_storage_test.exs
defmodule PersistentStorageTest do
  use ExUnit.Case
  # doctest PersistentStorage

  @test1_path "/tmp/persistent_storage_test_1"
  @test2_path "/tmp/persistent_storage_test_2"

  test "test basic storage behavior" do
    File.rm_rf! @test1_path
    File.rm_rf! @test2_path
    assert nil == PersistentStorage.get :test1, :this
    assert File.exists?(@test1_path) == false
    assert :ok = PersistentStorage.put :test1, this: "that"
    assert nil == PersistentStorage.get :test1, :someotherkey
    assert :foo == PersistentStorage.get :test1, :someotherkey, :foo
    assert "that" == PersistentStorage.get :test1, :this
    assert File.exists?(@test1_path) == true
    assert File.exists?(@test2_path) == false
    PersistentStorage.put :test2, test: {:a, 2}
    assert File.exists?(@test2_path) == true
    assert {:a, 2} == PersistentStorage.get :test2, :test
    PersistentStorage.put :test1, _weird_: %{strange: [{:value}]}
    assert File.exists? Path.join(@test1_path, "/this.storage")
    assert File.exists? Path.join(@test1_path, "/_weird_.storage")
    assert File.exists? Path.join(@test2_path, "/test.storage")
    assert "that" == PersistentStorage.get :test1, :this
    assert %{strange: [{:value}]} == PersistentStorage.get :test1, :_weird_
    assert %{strange: [{:value}]} == PersistentStorage.get :test1, :_weird_, :bogus_default
    PersistentStorage.put :test1, _weird_: %{strange: [{:value}]}
    assert %{strange: [{:value}]} == PersistentStorage.get :test1, :_weird_
    PersistentStorage.put :test1, _weird_: [:science]
    assert [:science] == PersistentStorage.get :test1, :_weird_
  end

end


mix.exs
defmodule PersistentStorage.Mixfile do
  use Mix.Project

  @version "0.10.1"

  def project do
    [app: :persistent_storage,
     version: @version,
     elixir: "~> 1.4",
     deps: deps(),
     description: "Brain dead simple file-based storage for embedded systems",
     package: package(),
     name: "PersistentStorage",
     docs: [
       source_ref: "v#{@version}", main: "PersistentStorage",
       source_url: "https://github.com/cellulose/persistent_storage",
#       main: "extra-readme",
       extras: [ "README.md", "CHANGELOG.md"] ]]
  end

  def application do
    [ mod: {PersistentStorage, []} ]
  end

  defp deps do
    [{:ex_doc, "~> 0.12", only: :dev}]
  end

  defp package do
    [ maintainers: ["Garth Hitchens", "Chris Dutton"],
      licenses: ["Apache-2.0"],
      links: %{github: "https://github.com/cellulose/persistent_storage"},
      files: ~w(lib config) ++ ~w(README.md CHANGELOG.md LICENSE mix.exs) ]
  end
end


lib/persistent_storage.ex
defmodule PersistentStorage do

  @moduledoc """
  Stores and retrieves terms from small flat files on embedded systems.

  `PersistentStorage` is intended for trivial persistent storage of basic system
  and application configuration information on embedded systems.   It is not
  intended to be a replacement for dets, sqlite, or many other far more capable
  databases.  It favors simplicity and robustness over performance and
  capability.

  ## Configuration

  Define one or more storage areas in your config.exs as follows...

  ```elixir
  # my_app/config/config.exs
  ...
  config :persistent_storage, tables: [
    settings: [path: "/root/storage/settings"],
    provisioning: [path: "/boot/provisioning"]
  ]
  ```

  The `:path` option is required is used to specify the filesystem path where
  files will be written.

  ## Usage

  ```elixir
  # write to settings (this will create the file at
  # /root/storage/settings/network
  iex> PersistentStorage.put :settings, :network, %{
    ip_address: {192,168,1,100}, mode: :static
  }

  # later, we can read it (this reads from ets cache if possible)
  iex> PersistentStorage.get :settings, :network
  %{ip_address: {192,168,1,100}, mode: :static}

  # read some provisioning data -- for purposes of this example, we
  # assume it was written when device  was first flashed)
  iex> PersistentStorage.get :provisioning, :device_data
  [serial_number: "302F1010", mfg_timestamp: "2016-05-04T03:28:35.279977Z"]
  ```
  """

  @type table :: atom
  @type key :: atom
  @type value :: any
  @type posix :: :file.posix

  # setup module attribute based on application env at compile time to allow
  # configuration of this module by applications that use it

  @tables Application.get_env :persistent_storage, :tables, []

  @doc """
  OTP Application start callback.

  Starts a supervisor with only one child process to own ETS tables.
  """
  @spec start(atom, term) :: {:ok, pid} | {:error, String.t}
  def start(_type, _args) do
    import Supervisor.Spec, warn: false
    children = [ worker(__MODULE__, []) ]
    Supervisor.start_link children, [strategy: :one_for_one, name: PersistentStorage.Supervisor]
  end

  @doc false
  @spec start_link() :: {:ok, pid()}
  def start_link do
    pid = spawn_link fn ->
      for {table, _opts} <- @tables do
        table_id = ets_table_for(table)
        ^table_id = :ets.new table_id, [ :set, :public, :named_table ]
      end
      receive do
        after :infinity -> :ok
      end
    end
    {:ok, pid}
  end

  @doc """
  Puts a single key and its associated value into storage.

  This is simple syntactic sugar for put/3 to allow specifying arguments
  with a keyword list consisting of a single key/value pair.

  ## Example

  ```elixir
  PersistentStorage.put :network_settings, router_ip_address: {192,168,15,1}
  ```
  """
  @spec put(table, [{key, value}]) :: :ok | {:error, posix}
  def put(table, [{key, value}]), do: put(table, key, value)

  @doc """
  Put a single key and its associated value into storage.

  ## Example

  ```elixir
  PersistentStorage.put :network_settings, :router_ip_address, {192,168,15,1}
  ```
  """
  @spec put(table, key, value) :: :ok | {:error, posix}
  def put(table, key, term) do
    :ets.insert ets_table_for(table), [{key, term}]
    case File.mkdir_p(directory_for(table)) do
      :ok ->
        File.write(path_for(table, key), :erlang.term_to_binary(term))
      error ->
        error
    end
  end

  @doc """
  Return the value associated with the given key in the specified storage.

  If the storage storage does not contain a key, returns the value of `default`
  (or nil if `default` is not provided).

  Uses the ets cache if possible, otherwise reads from disk.

  ## Examples

  ```elixir
  PersistentStorage.get :network_settings, :router_ip_address
  PersistentStorage.get :network_settings, :router_ip_address, {0,0,0,0}
  ```
  """
  @spec get(table, key, any) :: any
  def get(table, key, default_value \\ nil) do
    case :ets.lookup(ets_table_for(table), key) do
      [{^key, value}] ->
        value
      _ -> # not in ets cache
        case File.read(path_for(table, key)) do
          {:ok, contents} when (contents != "") ->
            term = :erlang.binary_to_term(contents)
            :ets.insert ets_table_for(table), [{key, term}]
            term
          _ ->
            default_value
        end
    end
  end

  @doc """
  Removes an entry from storage.

  ```elixir
  PersistentStorage.delete :network_settings, :router_ip_address
  ```
  """
  @spec delete(table, key) :: :ok
  def delete(table, key) do
    :ets.delete(ets_table_for(table), key)
    case File.rm(path_for(table, key)) do
      {:error, :enoent} ->
        :ok
      other ->
        other
    end
  end


  ### private helpers

  # returns the ets table for the specified storage
  defp ets_table_for(table), do: :"persistent_storage.#{table}"

  # return a directory given a storage name
  defp directory_for(table), do: @tables[table][:path]

  # given a storage and key, return a file path
  defp path_for(table, key) do
    Path.join directory_for(table), "#{key}.storage"
  end
end
